packages/dev/core/src/AudioV2/abstractAudio/subNodes/abstractAudioSubGraph.ts

```diff
@@ -29,23 +30,56 @@ export abstract class _AbstractAudioSubGraph {
      *
      * @param name The name of the sub node
      * @param callback The function to call with the named sub node
+     *
+     * @internal
      */
-    public callOnSubNode<T extends _AbstractAudioSubNode>(name: string, callback: (node: T) => void): void {
+    public callOnSubNode<T extends _AbstractAudioSubNode>(name: AudioSubNode, callback: (node: T) => void): void {
         const node = this.getSubNode(name);
         if (node) {
             callback(node as T);
             return;
         }
 
-        const promise = this._createSubNodePromises[name] ?? this._createAndAddSubNode(name);
+        const promise = this._createSubNodePromises[name] ?? this.createAndAddSubNode(name);
 
         promise.then((node) => {
             callback(node as T);
         });
     }
 
+    /**
+     * Creates the named subnode and adds it to the sub graph.
+     *
+     * @param name The name of the sub node.
+     * @returns A promise that resolves to the created sub node.
+     *
+     * @internal
+     */
+    public createAndAddSubNode(name: AudioSubNode): Promise<_AbstractAudioSubNode> {
+        const promise = this._createSubNode(name);
+
+        if (!promise) {
+            return Promise.reject(`Failed to create subnode "${name}"`);
+        }
+
+        this._createSubNodePromises[name] = new Promise((resolve, reject) => {
```

Comment:
if you already start from a promise, there is no need to have a resolve/reject, you can just do the promise.then and continue from there.
A better way would be to convert the function to async, and use async/await with try/catch

Reply:
yep, this seems perfectly fine. The `delete ...` part was not part of the original function, but I assume it is needed.

Also, no need to return the promise, you can just return the node. will be more readable.

you could have awaited on the first line (with try/catch to throw the error), but if you need the promise it's a different story.

Reply:
Hmm, looking at this more, I think if this change is made then the promise set on `_createSubNodePromises` will resolve before the call to `_addSubNode`, which will cause issues.

Reply:
Is there a way to make the `_createSubNode` promise stay unresolved until `_addSubNode` has been called? ...or do I need to go back to the original implementation that created a new Promise?

Reply:
Sorry, just realized you are deleted the promise you are assigning. Is the promise needed? if it isn't:

```javascript
    public async createAndAddSubNode(name: AudioSubNode): Promise<_AbstractAudioSubNode> {
        try {
           const node = await this._createSubNode(name);
           this._addSubNode(node);
           return node;
       } catch (e) {
          // alternatively avoid this catch and let it throw in the first await, making the stacktrace better)
           throw new Error(`Failed to create subnode "${name}"`);
      }
    }
```

Reply:
> Is there a way to make the `_createSubNode` promise stay unresolved until `_addSubNode` has been called? ...or do I need to go back to the original implementation that created a new Promise?

i'll probably need to understand why it is needed, but anything is possible. You dn't have to await the createSubNode, but if the node itself comes from the create function you will have to await it, so it must be resolved.

Reply:
The idea with `_createSubNode` is that it returns `null` if `name` doesn't match a known subnode. I did it this way so subclasses know whether a superclass created the node, or not. If I stick with that idea, then this `createAndAddSubNode` function is the earliest it can be known that the `name` parameter is invalid.

Reply:
Ok, I think I understand what to do now. Instead of returning the promise from `_createSubNode`, the cached promise is what needs to be returned after `_addSubNode` has been called. Like this:

```TypeScript
    public createAndAddSubNode(name: AudioSubNode): Promise<_AbstractAudioSubNode> {
        this._createSubNodePromises[name] = new Promise((resolve) => {
            const promise = this._createSubNode(name);

            if (!promise) {
                throw new Error(`Failed to create subnode "${name}"`);
            }

            promise.then((node) => {
                this._addSubNode(node);
                resolve(node);
            });
        });

        return this._createSubNodePromises[name];
    }
```

Reply:
> TBH - I am not sure what difference between this and throwing like this:
> 
> ```js
>     public async createAndAddSubNode(name: AudioSubNode): Promise<_AbstractAudioSubNode> {
>            const node = await this._createSubNode(name);
>           if(!node) { throw new Error(...); }
>            this._addSubNode(node);
>            return node;
>     }
> ```
> 
> But since it changes internal state I guess i don't see the whole picture

The idea for this function is it can be called once and awaited, or it can be called multiple times and the `_createSubNodePromisesResolved` function can be awaited instead. I think it makes sense to do it this way, but maybe it doesn't?

Reply:
Discussed offline and reworked things in commit https://github.com/BabylonJS/Babylon.js/pull/16260/commits/9b986be0851c1975a88dea5ad5e92f27e3c34f54.